crates/exex/src/manager.rs

```diff
@@ -258,75 +260,114 @@ impl Future for ExExManager {
     type Output = eyre::Result<()>;
 
     fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-        // drain handle notifications
-        while self.buffer.len() < self.max_capacity {
-            if let Poll::Ready(Some(notification)) = self.handle_rx.poll_recv(cx) {
-                debug!("received new notification");
-                self.push_notification(notification);
-                continue
+        loop {
+            // drain handle notifications
+            while self.buffer.len() < self.max_capacity {
+                if let Poll::Ready(Some(notification)) = self.handle_rx.poll_recv(cx) {
+                    debug!("received new notification");
+                    self.push_notification(notification);
+                    continue
+                }
+                break
             }
-            break
-        }
 
-        // update capacity
-        self.update_capacity();
-
-        // advance all poll senders
-        let mut min_id = usize::MAX;
-        for idx in (0..self.exex_handles.len()).rev() {
-            let mut exex = self.exex_handles.swap_remove(idx);
-
-            // it is a logic error for this to ever underflow since the manager manages the
-            // notification IDs
-            let notification_id = exex
-                .next_notification_id
-                .checked_sub(self.min_id)
-                .expect("exex expected notification ID outside the manager's range");
-            if let Some(notification) = self.buffer.get(notification_id) {
-                debug!(exex.id, notification_id, "sent notification to exex");
-                if let Poll::Ready(Err(err)) = exex.send(cx, notification) {
-                    // the channel was closed, which is irrecoverable for the manager
-                    return Poll::Ready(Err(err.into()))
+            // update capacity
+            self.update_capacity();
+
+            // This keeps track of the lowest notification ID that has been delivered to all ExExs.
+            let mut min_notification_id = usize::MAX;
+
+            // advance all exexs and send them notifications
+            for idx in (0..self.exex_handles.len()).rev() {
+                let mut exex = self.exex_handles.swap_remove(idx);
+
+            // ensure the exex has capacity
+            let mut reserve_result = exex.poll_reserve(cx);
+
+            // drain notifications
+            loop {
+                match reserve_result {
+                    Poll::Ready(Ok(())) => {
+                        // try send next notification to exex if we have it
+                        let notification_id = exex.next_notification_id;
```

Comment:
this is incorrect, you need to restore the previous logic for getting the notification id - the notification id is not a 1:1 index into the buffer

Reply:
Makes sense. I changed this so we also don't send the notification twice, and moved previous logic from outside the loop in here.